File: src/sfqc-jni/src/SFQueueClient.java
package jp.co.iret.sfq;

import java.util.Map;

public abstract class SFQueueClient implements SFQueueClientInterface
{
	static final int SFQ_RC_SUCCESS		= 0;
	static final int SFQ_RC_FATAL_MIN	= 21;

	int lastError_ = 0;
	String lastMessage_ = "";

	public void setLastError(int arg)
	{
		lastError_ = arg;
	}

	public int lastError()
	{
		return lastError_;
	}

	public void setLastMessage(final String arg)
	{
		lastMessage_ = arg;
	}

	public String lastMessage()
	{
		return lastMessage_;
	}

	public void clearLastError()
	{
		lastError_ = 0;
		lastMessage_ = "";
	}

	@Override
	public Map<String, Object> pop() throws SFQueueClientException
	{
		return pop(null);
	}
	@Override
	public Map<String, Object> shift() throws SFQueueClientException
	{
		return shift(null);
	}
}

File: src/sfq-cs/sfq/SFQueueClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace sfq
{
    using ObjectMap = Dictionary<String, Object>;

    abstract class SFQueueClient : SFQueueClientInterface
    {
        public abstract String push(ObjectMap param);
        public abstract ObjectMap pop(ObjectMap param);
        public abstract ObjectMap shift(ObjectMap param);

        protected const int SFQ_RC_SUCCESS   = 0;
        protected const int SFQ_RC_FATAL_MIN = 21;

        protected int LastError { set; get; }
        protected String LastMessage { set; get; }

        protected void clearLastError()
        {
            LastError = 0;
            LastMessage = "";
        }

        public ObjectMap pop()
        {
            return pop(new ObjectMap());
        }

        public ObjectMap shift()
        {
            return shift(new ObjectMap());
        }
    }
}
